pkg/github/token.go

```diff
@@ -0,0 +1,123 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gofri/go-github-ratelimit/github_ratelimit"
+	"github.com/google/go-github/v41/github"
+	"golang.org/x/oauth2"
+)
+
+type GitHubToken interface {
+	GetNewGitHubClient() (*github.Client, error)
+}
+
+type GitHubTokenClient struct {
+}
+
+// Tokens is mapping of token names to GitHub tokens
+var Tokens map[string]string
+
+// ParseGitHubTokens parses all of the possible GitHub tokens available to HAS and makes them available within the "github" package
+// This function should *only* be called once: at operator startup.
+func ParseGitHubTokens() error {
+	githubToken := os.Getenv("GITHUB_AUTH_TOKEN")
+	githubTokenList := os.Getenv("GITHUB_TOKEN_LIST")
+	if githubToken == "" && githubTokenList == "" {
+		return fmt.Errorf("no GitHub tokens were provided. Either GITHUB_TOKENS or GITHUB_AUTH_TOKEN (legacy) must be set")
+	}
+
+	Tokens = make(map[string]string)
+	if githubToken != "" {
+		// The old token format, stored in 'GITHUB_AUTH_TOKEN', didn't require a key/'name' for the token
+		// So use the key 'GITHUB_AUTH_TOKEN' for it
+		Tokens["GITHUB_AUTH_TOKEN"] = githubToken
+	}
+
+	// Parse any tokens passed in through the 'GITHUB_TOKEN_LIST' environment variable
+	// e.g. GITHUB_TOKEN_LIST=token1:ghp_faketoken,token2:ghp_anothertoken
+	if githubTokenList != "" {
+		// Each token key-value pair is separated by a comma, so split the string based on commas and loop over each key-value pair
+		tokenKeyValuePairs := strings.Split(githubTokenList, ",")
+		for _, tokenKeyValuePair := range tokenKeyValuePairs {
+			// Each token key-value pair is separated by a colon, so split the key-value pair and
+			// If the key-value pair doesn't split cleanly (i.e. only two strings returned), return an error
+			// If the key has already been added, return an error
+			splitTokenKeyValuePair := strings.Split(tokenKeyValuePair, ":")
+			if len(splitTokenKeyValuePair) != 2 {
+				return fmt.Errorf("unable to parse github token from key-value pair. Please ensure the GitHub secret is formatted correctly according to the documentation")
+			}
+			tokenKey := splitTokenKeyValuePair[0]
+			tokenValue := splitTokenKeyValuePair[1]
+
+			if Tokens[tokenKey] != "" {
+				return fmt.Errorf("a token with the key '%s' already exists. Each token must have a unique key", tokenKey)
+			}
+			Tokens[tokenKey] = tokenValue
+		}
+	}
+
+	return nil
+}
+
+// getRandomToken randomly retrieves a token from all of the tokens available to HAS
+func getRandomToken() (string, error) {
+	if len(Tokens) == 0 {
+		return "", fmt.Errorf("no GitHub tokens initialized")
+	}
+	var index int
+	if len(Tokens) == 1 {
+		index = 0
+	} else {
+		index = rand.Intn(len(Tokens) - 1)
```

Comment:
I don't think we need to subtract 1 since [rand.Intn](https://pkg.go.dev/math/rand#Intn) takes care of that by selecting a number between [0, n).  If `Tokens` is length 2, then 0 and 1 would be returned.  If we subtract 1 then 0 i.e. the first token would always be returned.   Having said that, I guess we don't need to check for length 1 and just have the statement `index=rand.Intn(len(Tokens))` take care of it all

Reply:
Yes, that's a lot better. Dunno why I didn't think of that haha.

Updated.